File: Frontend/src/DTOs/Pattern.ts
import { NeedleworkType } from "./Enums"
import { Notes } from "./Notes";
import { MyFile } from "./MyFile";
import { Yarn } from "./Yarn"
import { User } from "./User";
import { CountersGroup } from "./Counter";

interface PatternBase {
    id?: string;
    name: string;
    type: NeedleworkType;
    authorId: string;
    author?: User;
    category: string;
    notes: Notes[];
    photos?: MyFile[];
    files?: MyFile[];
    yarns: Yarn[];
    tools: Tool[];
    counters?: CountersGroup[];
    otherSupplies?: OtherSupply[];
    isShared: boolean;
}

export interface Pattern extends PatternBase {
    isAuthorial: boolean;
    startDate: string;
    endDate?: string;
    finished: boolean;
}

export interface CommunityPattern extends PatternBase {
    savedByUsers: User[];
}

export interface Tool {
    id?: string;
    name: string;
    quantity: number;
    size: string;
    patternId?: string;
}

export interface OtherSupply {
    id?: string;
    name: string;
    quantity: number;
    note?: string;
    patternId?: string;
}

File: Backend/src/Domain/Entities/Pattern.cs
using Common.Enums;

namespace Domain.Entities
{
    public class Pattern
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
        public NeedleworkType Type { get; set; }
        public string Category { get; set; }
        public string Notes { get; set; }
        public ICollection<Photo>? Photos { get; set; }
        public Guid AuthorId { get; set; }
        public User Author { get; set; }
        public ICollection<Yarn>? Yarns { get; set; }
        public ICollection<Tool>? Tools { get; set; }
        public ICollection<OtherSupply>? OtherSupplies { get; set; }
        public bool IsShared { get; set; }
        public ICollection<CountersGroup>? Counters { get; set; }
        public bool IsAuthorial { get; set; }
        public bool IsPlanned { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public ICollection<MyFile>? Files { get; set; }
        public bool IsFinished { get; set; }
        public ICollection<UserSavedCommunityPattern>? SavedByUsers { get; set; }


    }
}
